sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.nexmark.model.sql.adapter;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.beam.sdk.nexmark.model.Auction;
+import org.apache.beam.sdk.nexmark.model.Bid;
+import org.apache.beam.sdk.nexmark.model.Person;
+
+/**
+ * Maps Java model classes to Beam SQL record types.
+ */
+public class ModelFieldsAdapters {
```

Comment:
<!--thread_id:cc_151126021_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Generally speaking, I'm not a big fan of `ModelFieldsAdapter` and `ModelFieldsAdapters` classes. I understand that you have concrete adapters for nexmark model objects and `ModelFieldsAdapters.ADAPTERS` singleton to have all the adapters ready for use to convert in `ToBeamRecord` and also to get the coders. But IMHO I think it would be simpler and more object oriented to have the mapper/converter methods directly in Bid/Auction/Person... classes. WDYT? Is there something that I miss that prevents it?

<!--section:body-->
I see how it can be confusing. My reasonsing went something like this:

 - separation of concerns / data from algorithms. Stream of events produced by Generator is just data which you can apply algorithms on top of. Specific queries are implementation details of such algorithms. They are not limited to SQL. It's only current implementation details of SQL queries that requires objects to be converted to BeamRecords. So it feels that objects themselves should not be exposed to the fact that they are used by a specific kind of algorithms. For one thing, nothing stops Beam SQL to eventually be able to query objects directly without explicit in-advance conversion;

 - less coupling in general. Current Java models and generators have nothing to do with SQL. They can be used unchanged in any Beam-related code. For example I can see something like this used for integration testing. SQL representation of the models, on the other hand, can, and likely will change;

 - single responsibility. Current Java models are basically just serializable pojos. These utility classes also do one thing each. I see no reason to add behavior to pojos;

 - from what it looks like, not all models require conversion for use in SQL queries, only the ones being produced by the Generator. Query outputs are fine being BeamRecords, from what I can see so far;

 - my hope is that eventually this conversion layer will be gotten rid of, so that data is generated in a manner directly consumable by SQL. For example by making SQL understand Pojos, or by generating BeamRecords directly;


I hope this makes sense.

Reply:
Yes it makes a lot of sense, your aim is much clearer now. I agree with the low coupling and with the SRP too (even if there is not a lot of SRP in beam ;) ). Also, considering that the SQL converters might become useless in the future, extracting that conversion logic is a good thing indeed. The only conversion methods that belong to the model objects will be the encode/decode that are very unlikely to disappear in the future :)